service/history/nDCActivityReplicator.go

```diff
@@ -202,63 +202,70 @@ func (r *nDCActivityReplicatorImpl) shouldApplySyncActivity(
 		if err != nil {
 			return false, err
 		}
+
 		lastLocalItem, err := currentVersionHistory.GetLastItem()
 		if err != nil {
 			return false, err
 		}
+
 		incomingVersionHistory := persistence.NewVersionHistoryFromThrift(incomingRawVersionHistory)
 		lastIncomingItem, err := incomingVersionHistory.GetLastItem()
 		if err != nil {
 			return false, err
 		}
-		if lastIncomingItem.GetVersion() < lastLocalItem.GetVersion() {
-			// the incoming branch will lose to the local branch
-			// discard this task
-			return false, nil
-		}
 
 		lcaItem, err := currentVersionHistory.FindLCAItem(incomingVersionHistory)
 		if err != nil {
 			return false, err
 		}
 
-		// version history matches and activity schedule ID appears in local version history
-		if currentVersionHistory.IsLCAAppendable(lcaItem) && scheduleID <= lastLocalItem.GetEventID() {
-			return true, nil
-		}
+		if currentVersionHistory.IsLCAAppendable(lcaItem) || incomingVersionHistory.IsLCAAppendable(lcaItem) {
+			// case 1: local version history is superset of incoming version history
+			// or incoming version history is superset of local version history
+			if scheduleID > lcaItem.GetEventID() {
+				// get event from the incoming cluster
+				var endEventID *int64
+				var endEventVersion *int64
+				if scheduleID < lastIncomingItem.GetEventID() {
+					// resend is exclusive-exclusive. Get the last event to be the schedule event
+					endEventID = common.Int64Ptr(scheduleID + 1)
+					endVersion, err := incomingVersionHistory.GetEventVersion(scheduleID + 1)
+					if err != nil {
+						return false, err
+					}
+					endEventVersion = common.Int64Ptr(endVersion)
+				}
```

Comment:
plz remove this section, just resend to the end of history

Reply:
removed